### PR TITLE
Fix an early-halt condition, misc refactorings

### PR DESCRIPTION
**Pull request by Mats-SX**
This class implements quite complex functionality, so I have opted for more descriptive (longer) names and made an effort to clarify that many values are estimates rather than exact statistics.

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/cardinality/assumeIndependence/PatternSelectivityCalculator.scala [hunk 1]
```diff
@@ -42,24 +42,26 @@ case class PatternSelectivityCalculator(stats: GraphStatistics, combiner: Select
...
-    val lhsCardinality = allNodes * calculateLabelSelectivity(labelsOnLhs)
-    val rhsCardinality = allNodes * calculateLabelSelectivity(labelsOnRhs)
+    val lhsCardinality = nbrOfNodesInGraph * calculateLabelSelectivity(labelsOnLhs, nbrOfNodesInGraph)
+    val rhsCardinality = nbrOfNodesInGraph * calculateLabelSelectivity(labelsOnRhs, nbrOfNodesInGraph)
 
     // If either side of our pattern is empty, it's all empty
-    if (lhsCardinality == Cardinality.EMPTY || lhsCardinality == Cardinality.EMPTY)
-      Selectivity(1)
+    if (lhsCardinality == Cardinality.EMPTY || rhsCardinality == Cardinality.EMPTY)
+      Selectivity.ZERO
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/cardinality/assumeIndependence/PatternSelectivityCalculator.scala:56]: This was a minor bug in the if-clause, but surely the selectivity factor should be zero and not one, in the case where no endpoint nodes (for some end) exist for the pattern?

Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/cardinality/assumeIndependence/PatternSelectivityCalculator.scala:56]: Having 0 selectivity is usually very wrong. Mind that selectivity is not a cardinality. What's wrong in giving back 1?

Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/cardinality/assumeIndependence/PatternSelectivityCalculator.scala:56]: Yes, selectivity is not cardinality. But the _value_ for the selectivity in our code is directly proportional to cardinality. This means that selectivity zero => cardinality zero. If a pattern has no possible endpoint (for one of the endpoints), then surely the cardinality of the pattern should be zero, right?

`MATCH (a:A)-->(b:B)` and there are no `:B`-nodes, then surely there can be no matches.

Comment (boggle) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/cardinality/assumeIndependence/PatternSelectivityCalculator.scala:56]: I can buy that but @davidegrohmann point is valid, too. Let's merge and see if it trips up in iscyfayet. Perhaps we should have some Selectivity.EPSILON instead?

**Pull request merged**